### PR TITLE
Next firefox clipboard

### PR DESCRIPTION
**Pull request by davidfig**
- [x] better support for Firefox copy/pasting
- [x] full keyboard shortcut clipboard support for firefox (with fallback for menu copying)

### CODE REVIEW

File: src/grid/actions/clipboard/clipboard.ts [hunk 1]
```diff
@@ -1,22 +1,105 @@
...
+import { pixiAppSettings } from '../../../gridGL/pixiApp/PixiAppSettings';
 import { copyAsPNG } from '../../../gridGL/pixiApp/copyAsPNG';
-import { Coordinate } from '../../../gridGL/types/size';
 import { grid } from '../../controller/Grid';
 import { sheets } from '../../controller/Sheets';
 
+//#region document event handler for copy, paste, and cut
+
+export const copyToClipboardEvent = (e: ClipboardEvent) => {
+  debugTimeReset();
```
Comment (davidkircos) [inline src/grid/actions/clipboard/clipboard.ts:11]: Did you mean to leave these debug function calls in?

File: src/grid/actions/clipboard/clipboard.ts [hunk 2]
```diff
@@ -1,22 +1,105 @@
...
+  const rectangle = sheets.sheet.cursor.getRectangle();
+  const { plainText, html } = grid.copyToClipboard(sheets.sheet.id, rectangle);
+  if (!e.clipboardData) {
+    console.warn('clipboardData is not defined');
```
Comment (davidkircos) [inline src/grid/actions/clipboard/clipboard.ts:15]: When would this happen? Do we want to report to sentry or is this sometimes expected?

File: src/gridGL/interaction/keyboard/keyboardClipboard.ts [hunk 3]
```diff
@@ -29,34 +16,5 @@ export function keyboardClipboard(props: {
...
-  // Don't allow commands past here without permission
-  if (!isEditorOrAbove(permission)) {
-    return false;
-  }
-
-  // Command + X
-  if ((event.metaKey || event.ctrlKey) && event.key === 'x') {
-    cutToClipboard(start, end);
-    return true;
-  }
```
Comment (davidkircos) [inline src/gridGL/interaction/keyboard/keyboardClipboard.ts:53]: cut no longer works on Chrome on Mac via Command + X

File: src/grid/actions/clipboard/clipboard.ts [hunk 4]
```diff
@@ -30,66 +113,80 @@ const toClipboard = (plainText: string, html: string) => {
...
-        html,
-      });
-      debugTimeCheck('paste from clipboard');
-    } catch (e) {
-      console.warn(e);
+export const pasteFromClipboard = async () => {
+  if (!isEditorOrAbove(pixiAppSettings.permission)) return;
+  const target = sheets.sheet.cursor.originPosition;
+
+  if (navigator.clipboard?.read) {
```
Comment (davidkircos) [inline src/grid/actions/clipboard/clipboard.ts:154]: The clipboard logic here is a bit complex and platform-specific. Can you add some detailed comments so when we edit this in a year, we remember why we did what we did?

File: src/gridGL/pixiApp/PixiApp.ts [hunk 5]
```diff
@@ -127,10 +128,14 @@ export class PixiApp {
 
   private setupListeners() {
     window.addEventListener('resize', this.resize);
+    document.addEventListener('copy', copyToClipboardEvent);
+    document.addEventListener('paste', pasteFromClipboardEvent);
```
Comment (davidkircos) [inline src/gridGL/pixiApp/PixiApp.ts:132]: Probably forgot to add an event listener for `cut`

File: src/grid/actions/clipboard/clipboard.ts [hunk 6]
```diff
@@ -1,22 +1,105 @@
...
+
+  e.preventDefault();
+};
+
+//#regionend
+
+//#region triggered via menu (limited support on Firefox)
+
+// workaround so Firefox can copy/paste within same app
+let lastCopy: string | undefined = undefined;
```
Comment (davidkircos) [inline src/grid/actions/clipboard/clipboard.ts:78]: code style, define at the top of the file?

Comment (davidkircos) [reply src/grid/actions/clipboard/clipboard.ts:78]: Does it make sense to use local storage so you can copy/paste between tabs on Firefox?

File: src/grid/actions/clipboard/clipboard.ts [hunk 7]
```diff
@@ -30,66 +113,80 @@ const toClipboard = (plainText: string, html: string) => {
...
+  // todo: this does not work in firefox
   if (navigator.clipboard && window.ClipboardItem) {
     navigator.clipboard.write([
       new ClipboardItem({
         //@ts-ignore
         'image/png': blob,
       }),
     ]);
+  } else {
+    console.log('copy to PNG is not supported in Firefox (yet)');
```
Comment (davidkircos) [inline src/grid/actions/clipboard/clipboard.ts:146]: Ah, we should maybe hide this feature if it just doesn't work in FireFox. If that's the case I think it's beyond the scope of this PR and we should just create an issue for later.

**Review by davidkircos — CHANGES_REQUESTED** [review]

File: src/grid/actions/clipboard/clipboard.ts (hunk 1 above)
Comment (davidfig) [reply src/grid/actions/clipboard/clipboard.ts:11]: Yes. These only run when `debugShowTime = true`

File: src/grid/actions/clipboard/clipboard.ts (hunk 2 above)
Comment (davidfig) [reply src/grid/actions/clipboard/clipboard.ts:15]: This should not happen in any browser we support. But it's optional in the types.

File: src/grid/actions/clipboard/clipboard.ts [hunk 8]
```diff
@@ -1,22 +1,114 @@
...
+
+//#endregion
+
+//#region triggered via menu (limited support on Firefox)
+
 // copies plainText and html to the clipboard
 const toClipboard = (plainText: string, html: string) => {
   // https://github.com/tldraw/tldraw/blob/a85e80961dd6f99ccc717749993e10fa5066bc4d/packages/tldraw/src/state/TldrawApp.ts#L2189
-  if (navigator.clipboard && window.ClipboardItem) {
+  if (navigator.clipboard) {
```
Comment (davidkircos) [inline src/grid/actions/clipboard/clipboard.ts:92]: Should we use the `fullClipboardSupport ` function here?

File: src/grid/actions/clipboard/clipboard.ts [hunk 9]
```diff
@@ -30,66 +122,90 @@ const toClipboard = (plainText: string, html: string) => {
...
-      if (htmlItem) {
-        const item = await htmlItem.getType('text/html');
-        html = await item.text();
-      }
+export const pasteFromClipboard = async () => {
+  if (!isEditorOrAbove(pixiAppSettings.permission)) return;
+  const target = sheets.sheet.cursor.originPosition;
+
+  // handles non-Firefox browsers
+  if (navigator.clipboard?.read) {
```
Comment (davidkircos) [inline src/grid/actions/clipboard/clipboard.ts:164]: use `fullClipboardSupport`?

File: src/grid/actions/clipboard/clipboard.ts [hunk 10]
```diff
@@ -1,22 +1,109 @@
...
+      sheetId: sheets.sheet.id,
+      x: cursor.x,
+      y: cursor.y,
+      plainText,
+      html,
+    });
+  }
+
+  // enables Firefox menu pasting after a ctrl+v paste
+  localforage.setItem(clipboardLocalStorageKey, html);
```
[the remark this answers is not in the document]
Comment (davidfig) [reply src/grid/actions/clipboard/clipboard.ts:80]: This was how Figma handled it, so I thought it was helpful.

File: src/grid/actions/clipboard/clipboard.ts (hunk 8 above)
Comment (davidfig) [reply src/grid/actions/clipboard/clipboard.ts:92]: Sure

File: src/grid/actions/clipboard/clipboard.ts (hunk 9 above)
Comment (davidfig) [reply src/grid/actions/clipboard/clipboard.ts:147]: This one is slightly different depending on when and how Firefox decides to support clipboards.

**Review by davidkircos — APPROVED** [review]

**Pull request merged**